File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_4_C_Dictionary/ALDS1_4_C.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type hash struct {
	ss [1000000]string
}

func (h *hash) insert(data string) {
	k := toKey(data)
	i := 0
	hValue := h.hash(k, i)
	v := h.ss[hValue]
	for v != "" && v != data {
		i++
		hValue = h.hash(k, i)
		v = h.ss[hValue]
	}
	if h.ss[hValue] == "" {
		h.ss[hValue] = data
	}
}

func (h *hash) search(data string) bool {
	k := toKey(data)
	i := 0
	v := h.ss[h.hash(k, i)]
	for v != "" && v != data {
		i++
		v = h.ss[h.hash(k, i)]
	}
	if v == "" {
		return false
	}
	return true
}

func (h *hash) hash(k int, i int) int {
	return (h1(k) + i*h2(k)) % 1046527
}

func h1(k int) int {
	return k % 1046527
}

func h2(k int) int {
	return 1 + (k % 1046526)
}

func toKey(s string) int {
	result := 0
	for _, c := range s {
		switch c {
		case 'A':
			result = result + 1
		case 'C':
			result = result + 2
		case 'G':
			result = result + 3
		case 'T':
			result = result + 4
		default:
		}
	}
	return result
}

func nextInt(sc *bufio.Scanner) int {
	sc.Scan()
	n, err := strconv.Atoi(sc.Text())
	if err != nil {
		panic(err)
	}
	return n
}

func nextStr(sc *bufio.Scanner) string {
	sc.Scan()
	s := sc.Text()
	return s
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanWords)

	var h hash

	n := nextInt(sc)
	for n > 0 {
		cmd := nextStr(sc)
		v := nextStr(sc)
		switch cmd {
		case "insert":
			h.insert(v)
		case "find":
			if h.search(v) {
				fmt.Println("yes")
			} else {
				fmt.Println("no")
			}
		default:
		}
		n--
	}
}

File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_4_C_Dictionary/ALDS1_4_C.rb
def to_key(s)
  s.gsub!('A', '1')
  s.gsub!('C', '2')
  s.gsub!('G', '3')
  s.gsub!('T', '4')
  s.to_i
end
def hash(key, i)
  (h1(key) + i * h2(key)) % 1046527
end
def h1(key)
  key % 1046527
end
def h2(key)
  1 + (key % (1046527 - 1))
end

class Dict
  def initialize
    @arr = []
  end
  def insert(s)
    key = to_key(s)
    i = 0
    loop do
      h = hash(key, i)
      if @arr[h] == s
        return
      elsif @arr[h].nil?
        @arr[h] = s
        return
      end
      i = i + 1
    end
  end
  def find(s)
    key = to_key(s)
    i = 0
    loop do
      h = hash(key, i)
      if @arr[h] == s
        return true
      elsif @arr[h].nil?
        return false
      end
      i = i + 1
    end
    false
  end
end

n = STDIN.gets.to_i

dict = Dict.new
n.times.each {
  ss = STDIN.gets.split(' ')
  if ss[0] == 'insert' then
    dict.insert(ss[1])
  else
    if dict.find(ss[1]) then
      puts 'yes'
    else
      puts 'no'
    end
  end
}

